lib/codegen/Decision.cpp

```diff
@@ -65,12 +65,21 @@ void SwitchNode::codegen(Decision *d, llvm::StringMap<llvm::Value *> substitutio
     }
     d->CurrentBlock = entry.first;
     if (!isInt) {
-      int offset = 2;
+      int offset = 0;
       llvm::StructType *BlockType = getBlockType(d->Module, d->Definition, _case.getConstructor());
       llvm::BitCastInst *Cast = new llvm::BitCastInst(val, llvm::PointerType::getUnqual(BlockType), "", d->CurrentBlock);
       for (std::string binding : _case.getBindings()) {
-        llvm::Value *ChildPtr = llvm::GetElementPtrInst::CreateInBounds(BlockType, Cast, {llvm::ConstantInt::get(llvm::Type::getInt64Ty(d->Ctx), 0), llvm::ConstantInt::get(llvm::Type::getInt32Ty(d->Ctx), offset++)}, "", d->CurrentBlock);
-        substitution[binding] = new llvm::LoadInst(ChildPtr, binding, d->CurrentBlock);
+        llvm::Value *ChildPtr = llvm::GetElementPtrInst::CreateInBounds(BlockType, Cast, {llvm::ConstantInt::get(llvm::Type::getInt64Ty(d->Ctx), 0), llvm::ConstantInt::get(llvm::Type::getInt32Ty(d->Ctx), offset+2)}, "", d->CurrentBlock);
+        switch (dynamic_cast<KOREObjectCompositeSort *>(_case.getConstructor()->getArguments()[offset++])->getCategory(d->Definition).cat) {
+        case SortCategory::Map:
+        case SortCategory::List:
+        case SortCategory::Set:
+          substitution[binding] = ChildPtr;
+          break;
+        default:
+          substitution[binding] = new llvm::LoadInst(ChildPtr, binding, d->CurrentBlock);
```

Comment:
Is this consistent with the function call above? It looks like this is making the substitution binding hold the address of the field within the parent term, rather than something that only references the collection.

Also, if so, pointers to the interior of objects can make garbage collection harder.

Reply:
it's consistent, yes. we pass map/list/set by reference (ie via a pointer) and everything else by value, hence the load.